### DISCUSSION TITLE
Is it possible to create files with template variable replacement?

### DISCUSSION THREAD

**Discussion by moltar** [renovatebot/renovate#26175] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

38

### If you're self-hosting Renovate, select which platform you are using.

github.com

### What is your question?

I am trying to solve the problem explained in #24882 in a slightly different way.

I want to use `postUpgradeTasks` option with templates to create a file for changesets.

Initially, I was thinking to just do `echo "template" > .changesets/foo.md`,  but then I realized that this would not work, as _after_ template replacement, there might be characters that will not be escaped and fail to echo.

So, I am wondering, is there an option to _create_ a new file, and pass it thru a templating engine?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment]
postUpgradeTasks can run anything you like, including templating engines. But you would need to make sure that any engine you need is installed

**Reply by moltar** [reply]
Sure, that is understood. But how I'd pass template values to the engine. The only way I'm aware of is the template substitution by Renovate itself. So it'd still be a shell command.

**Reply by rarkins** [reply]
How do you normally pass _values_ to the engine? e.g. via CLI params, via env, via file?

**Reply by moltar** [reply]
I normally do not do it.

But given that the only option would be to run in `postUpgradeTasks`, it has to be a CLI then?

I envision something like:

```json
"postUpgradeTasks": [
  "echo \"{{var}}\" > .changesets/changeset-foo.md"
]
```

But my understanding is that _before_ executing the command, Renovate will replace the templated variables, so that before the command is ran, it will be:

```json
"postUpgradeTasks": [
  "echo \"REPLACED VALUES HERE\" > .changesets/changeset-foo.md"
]
```

So my concern is that if `REPLACED VALUES HERE` contains any escape characters, the command will fail.

For example, imagine including the entire `{{body}}` (changelog content) there. I think it'd be a huge mess, and there's no way to escape it.

**Reply by rarkins** [reply]
I'm not sure there's any solution to the escaping characters problem here

**Reply by moltar** [reply]
One possible solution, which currently does not exist, but would be relatively straightforward to implement is a Handlebars helper to convert to Base64, similar to `encodeURIComponent`.

Which, in Node.js, is basically just a `Buffer.from(value).toString('base64')`

https://docs.renovatebot.com/templates/#additional-handlebars-helpers

Then:

```
echo "{{{base64 baseDir}}}" | base64 -d > foo.md
```

https://github.com/renovatebot/renovate/blob/7d69fa954e014b88a8d104e57ee6e4cee921c499/lib/util/template/index.ts#L8-L26

**Reply by rarkins** [reply, marked as answer]
That sounds like a good idea - PR welcome for adding that helper

**Reply by moltar** [reply]
https://github.com/renovatebot/renovate/pull/26197
